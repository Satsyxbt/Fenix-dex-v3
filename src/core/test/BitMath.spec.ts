import { expect } from './shared/expect';
import { BitMathTest } from '../typechain';
import { ethers } from 'hardhat';
import { loadFixture } from '@nomicfoundation/hardhat-network-helpers';
import snapshotGasCost from './shared/snapshotGasCost';

describe('BitMath', () => {
  let bitMath: BitMathTest;
  const fixture = async () => {
    const factory = await ethers.getContractFactory('BitMathTest');
    return (await factory.deploy()) as any as BitMathTest;
  };
  beforeEach('deploy BitMathTest', async () => {
    bitMath = await loadFixture(fixture);
  });

  describe('#leastSignificantBit', () => {
    it('1', async () => {
      expect(await bitMath.leastSignificantBit(1)).to.eq(0);
    });
    it('2', async () => {
      expect(await bitMath.leastSignificantBit(2)).to.eq(1);
    });
    it('all powers of 2', async () => {
      const results = await Promise.all(
        [...Array(255)].map((_, i) => bitMath.leastSignificantBit(2n ** BigInt(i)))
      );
      expect(results).to.deep.eq([...Array(255)].map((_, i) => i));
    });
    it('uint256(-1)', async () => {
      expect(await bitMath.leastSignificantBit(2n ** 256n - 1n)).to.eq(0);
    });

    it('gas cost of smaller number  [ @skip-on-coverage ]', async () => {
      await snapshotGasCost(bitMath.getGasCostOfLeastSignificantBit(3568n));
    });
    it('gas cost of min number  [ @skip-on-coverage ]', async () => {
      await snapshotGasCost(bitMath.getGasCostOfLeastSignificantBit(1n));
    });
    it('gas cost of max uint128  [ @skip-on-coverage ]', async () => {
      await snapshotGasCost(bitMath.getGasCostOfLeastSignificantBit(2n ** 128n - 1n));
    });
    it('gas cost of max uint256  [ @skip-on-coverage ]', async () => {
      await snapshotGasCost(bitMath.getGasCostOfLeastSignificantBit(2n ** 256n - 1n));
    });
  });
});
