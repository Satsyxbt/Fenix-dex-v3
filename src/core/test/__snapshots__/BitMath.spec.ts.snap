// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BitMath #leastSignificantBit gas cost of max uint128  [ @skip-on-coverage ] 1`] = `216`;

exports[`BitMath #leastSignificantBit gas cost of max uint256  [ @skip-on-coverage ] 1`] = `216`;

exports[`BitMath #leastSignificantBit gas cost of min number  [ @skip-on-coverage ] 1`] = `216`;

exports[`BitMath #leastSignificantBit gas cost of smaller number  [ @skip-on-coverage ] 1`] = `216`;
